src/Features/Core/Portable/Snippets/RoslynLSPSnippetConverter.cs

```diff
@@ -0,0 +1,174 @@
+                        i++;
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    lspSnippetString.Append(str);
+                    i += strLength;
+                }
+            }
+
+            return lspSnippetString.ToString();
+        }
+
+        /// <summary>
+        /// Preprocesses the list of placeholders into a dictionary that maps the insertion position
+        /// in the string to the placeholder's identifier and the number associated with it.
+        /// </summary>
+        private static Dictionary<int, (string identifier, int priority)> GetMapOfSpanStartsToLSPStringItem(ImmutableArray<SnippetPlaceholder> placeholders, int textChangeStart)
+        {
+            var map = new Dictionary<int, (string, int)>();
+
+            for (var i = 0; i < placeholders.Length; i++)
+            {
+                var placeholder = placeholders[i];
+                foreach (var position in placeholder.PlaceHolderPositions)
+                {
+                    map.Add(position - textChangeStart, (placeholder.Identifier, i + 1));
+                }
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Tries to see if a value exists at that position in the map, and if so it
+        /// generates a string that is LSP formatted as well as passes back the length
+        /// of the identifier so that it can skip forward in the string.
+        /// </summary>
+        private static (string str, int strLength) GetStringInPosition(Dictionary<int, (string identifier, int priority)> map, int position)
+        {
+            if (map.TryGetValue(position, out var placeholderInfo))
+            {
+                return ($"${{{placeholderInfo.priority}:{placeholderInfo.identifier}}}", placeholderInfo.identifier.Length);
+            }
+
+            return (string.Empty, 0);
+        }
+
+        /// <summary>
+        /// We need to extend the snippet's TextChange if any of the placeholders or
+        /// if the caret position comes before or after the span of the TextChange.
+        /// If so, then find the new string that encompasses all of the placeholders
+        /// and caret position.
+        /// </summary>
+        private static async Task<TextChange> ExtendSnippetTextChangeAsync(Document document, TextChange textChange, ImmutableArray<SnippetPlaceholder> placeholders, int caretPosition)
+        {
+            var extendedSpan = GetUpdatedTextSpan(textChange, placeholders, caretPosition);
+
+            if (extendedSpan.Length == 0)
+            {
+                return textChange;
+            }
+
+            var documentText = await document.GetTextAsync().ConfigureAwait(false);
+            var newString = documentText.ToString(extendedSpan);
+            var newTextChange = new TextChange(new TextSpan(extendedSpan.Start, 0), newString);
```

Comment:
i'm confused by this.  you extended a text range, but your new extended change has a span that has a length of 0.  that means the range collapsed.  that seems very strange.

Reply:
should the length be the newstring length?

Reply:
i don't know.  you tell me :)   but i don't understand the logic here that you're starting with a text change (representing a span to replace in the original doc, then you're extending it, but you end up with no length to replace once you extend.

Reply:
still feel like there is somerthing wonkky here.